pkg/zng/docs/spec.md

```diff
@@ -1,92 +1,366 @@
+comprised of a sequence of newline-delimited UTF-8 strings,
+as well as a binary form called "BZNG", which is based on machine-readable data types
+inspired by [Protobufs](https://developers.google.com/protocol-buffers).
+
+ZNG is richly typed and thinner on the wire than JSON.
 Like [newline-delimited JSON (NDJSON)](http://ndjson.org/),
-ZSON represents a sequence of data objects that can be parsed line by line.
+the ZNG text format represents a sequence of data objects
+that can be parsed line by line.
 
-ZSON strikes a balance between the narrowly typed but flexible NDJSON format and
+ZNG strikes a balance between the narrowly typed but flexible NDJSON format and
 a more structured approach like
 [Apache Avro](https://avro.apache.org).
-ZSON is type rich and
-embeds all type/schema in the stream, while having a value syntax
-independent of the schema, making it easy and efficient to parse on the fly
-and mix and match streams from different sources with heterogeneous types.
-Like Avro,
-ZSON embeds schema information in the data stream but ZSON schema bindings
-are lighter-weight and specified with a simple integer encoding that
-accompanies each data value.
-
-The ZSON design [is also motivated by](./rationale.md)
+ZNG is type rich and embeds all type/schema in the stream while having a
+binary serialization format that allows "lazy parsing" of fields such that
+only the fields of interest in a stream need to be extracted and parsed.
+Unlike Avro and Protobufs,
+ZNG embeds type information in the data stream and admits efficient
+multiplexing of heterogenous data types by including with each data value
+a simple integer identifier to reference its type.
+
+ZNG is a superset of JSON in that any JSON input
+can be mapped onto ZNG and recovered by decoding
+that ZNG back into JSON.
+
+The ZNG design [is also motivated by](./rationale.md)
 and maintains backward compatibility with the original [default Zeek log format](https://docs.zeek.org/en/stable/examples/logs/).
 
-## ZSON format
+## The ZNG data model
+
+ZNG encodes a sequence of one or more typed data values comprising a stream.
+The stream of values is interleaved with control messages
+that provide type definitions for the data values.  The type of
+a particular data value is specified by its "type code", which is an
+integer identifier representing either a built-in type or
+or a composite type definition occurring previously in the stream.
+
+ZNG is designed for efficient use as a protocol for streaming values between
+end systems and thus allows control messages to carry arbitrary data payloads as
+signaling from a higher-layer protocol that is embedded in the ZNG stream.
+
+The ZNG type system comprises the standard set
+of types like integers, floating point,
+strings, byte arrays, etc as well as composite types build from the
+standard types including records, arrays, and sets.
+
+For example, a ZNG stream representing the single string "hello world" conceptually
```

Comment:
I was trying to define the model followed by the details but it didn't work out very well.  Rather than try to improve this section, maybe I could just drop it and have a couple sentences about the model.

Reply:
Ok, I simplified this section a bit a rewrote the example in ZNG syntax.